resource-aggregate/cqrs/eventstore/projection.go

```diff
@@ -289,44 +289,70 @@ func (p *Projection) Forget(queries []SnapshotQuery) (err error) {
 	return nil
 }
 
-func (p *Projection) allModelsLocked(onModel func(m Model) (wantNext bool)) {
+func (p *Projection) iterateOverAllModels(onModel func(m Model) (wantNext bool)) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
 	for _, group := range p.aggregateModels {
 		for _, apm := range group {
-			if !onModel(apm.model) {
-				return // stop iteration
+			p.lock.RUnlock()
```

Comment:
This works, but instead of locking and unlocking in each step wouldn't something like this be better:

```
p.lock.RLock()
create deep copy of p.aggregateModels
p.lock.RUnlock()

iterate over the copy
```

Reply:
This is what I want to avoid.  `aggregateModels` can contains 500K elements so copy is to expensive on CPU.

Reply:
mainly allocation